File: packages/client-sdk/src/services/api-keys.ts
/**
 * API Keys service
 */

import type { HttpClient } from '../core/http';
import type { MessageResponse } from '../types/common';
import type {
  APIKey,
  CreateAPIKeyRequest,
  CreateAPIKeyResponse,
  ListAPIKeysResponse,
  UpdateAPIKeyRequest,
} from '../types/api-key';
import { BaseService } from './base';

/** API Keys service for managing API keys */
export class APIKeysService extends BaseService {
  constructor(http: HttpClient) {
    super(http);
  }

  /**
   * Create a new API key
   * @param data API key creation data
   * @returns Created API key with plain key (only shown once!)
   */
  async create(data: CreateAPIKeyRequest): Promise<CreateAPIKeyResponse> {
    const response = await this.http.post<CreateAPIKeyResponse>(
      '/api/api-keys',
      data
    );
    return response.data;
  }

  /**
   * List all API keys for the current user
   * @returns List of API keys (without plain key values)
   */
  async list(): Promise<ListAPIKeysResponse> {
    const response = await this.http.get<ListAPIKeysResponse>('/api/api-keys');
    return response.data;
  }

  /**
   * Get a specific API key by ID
   * @param id API key ID
   * @returns API key details
   */
  async get(id: string): Promise<APIKey> {
    const response = await this.http.get<APIKey>(`/api-keys/${id}`);
    return response.data;
  }

  /**
   * Update an API key
   * @param id API key ID
   * @param data Update data
   * @returns Updated API key
   */
  async update(id: string, data: UpdateAPIKeyRequest): Promise<APIKey> {
    const response = await this.http.put<APIKey>(`/api-keys/${id}`, data);
    return response.data;
  }

  /**
   * Revoke an API key (deactivate but keep record)
   * @param id API key ID
   * @returns Success message
   */
  async revoke(id: string): Promise<MessageResponse> {
    const response = await this.http.post<MessageResponse>(
      `/api-keys/${id}/revoke`
    );
    return response.data;
  }

  /**
   * Delete an API key permanently
   * @param id API key ID
   * @returns Success message
   */
  async delete(id: string): Promise<MessageResponse> {
    const response = await this.http.delete<MessageResponse>(`/api-keys/${id}`);
    return response.data;
  }

  /**
   * Get all active API keys
   * @returns List of active API keys
   */
  async getActive(): Promise<APIKey[]> {
    const { api_keys } = await this.list();
    return api_keys.filter((key: APIKey) => key.is_active);
  }
}

File: packages/go-sdk/apikeys.go
package authgateway

import (
	"context"
	"fmt"

	"github.com/smilemakc/auth-gateway/packages/go-sdk/models"
)

// APIKeysService handles API key operations.
type APIKeysService struct {
	client *Client
}

// Create creates a new API key.
// The plain key is only returned once at creation - store it securely.
func (s *APIKeysService) Create(ctx context.Context, req *models.CreateAPIKeyRequest) (*models.CreateAPIKeyResponse, error) {
	var resp models.CreateAPIKeyResponse
	if err := s.client.post(ctx, "/api/api-keys", req, &resp); err != nil {
		return nil, err
	}
	return &resp, nil
}

// List retrieves all API keys for the current user.
func (s *APIKeysService) List(ctx context.Context) ([]models.APIKey, error) {
	var resp []models.APIKey
	if err := s.client.get(ctx, "/api/api-keys", &resp); err != nil {
		return nil, err
	}
	return resp, nil
}

// Get retrieves a specific API key by ID.
func (s *APIKeysService) Get(ctx context.Context, id string) (*models.APIKey, error) {
	var resp models.APIKey
	if err := s.client.get(ctx, fmt.Sprintf("/api/api-keys/%s", id), &resp); err != nil {
		return nil, err
	}
	return &resp, nil
}

// Update updates an API key.
func (s *APIKeysService) Update(ctx context.Context, id string, req *models.UpdateAPIKeyRequest) (*models.APIKey, error) {
	var resp models.APIKey
	if err := s.client.put(ctx, fmt.Sprintf("/api/api-keys/%s", id), req, &resp); err != nil {
		return nil, err
	}
	return &resp, nil
}

// Revoke revokes an API key (deactivates it).
func (s *APIKeysService) Revoke(ctx context.Context, id string) (*models.MessageResponse, error) {
	var resp models.MessageResponse
	if err := s.client.post(ctx, fmt.Sprintf("/api/api-keys/%s/revoke", id), nil, &resp); err != nil {
		return nil, err
	}
	return &resp, nil
}

// Delete permanently deletes an API key.
func (s *APIKeysService) Delete(ctx context.Context, id string) error {
	return s.client.delete(ctx, fmt.Sprintf("/api/api-keys/%s", id), nil)
}
